### ISSUE TITLE
Prevent memory overflow with initial loading datasplit

### ISSUE THREAD

**Issue by rhoadesScholar** [janelia-cellmap/cellmap-segmentation-challenge#98]
> Gotcha. Thank you! 
> @paperplane03 I will add a feature request for limiting the datasplit creation to prevent RAM overflow. I go on paternity leave next week, so I am not sure when it will get addressed though. 

 _Originally posted by @rhoadesScholar in [#82](https://github.com/janelia-cellmap/cellmap-segmentation-challenge/issues/82#issuecomment-2648824198) replying to (https://github.com/janelia-cellmap/cellmap-segmentation-challenge/issues/82#issuecomment-2644476291)

**Comment by paperplane03** [comment]
Thank you!

**Comment by paperplane03** [comment]
I find the problem is related the to scale and I roughly know how to locate the problem. 
Can I know what is the output if you add this two lines in train.py:220?
`lengthes = [len(x) for x in train_loader.dataset.datasets]`
`print(lengthes)`

**Comment by ddd9898** [comment]
Hi @paperplane03 , I'm having the same problem as you mentioned in https://github.com/janelia-cellmap/cellmap-segmentation-challenge/issues/82#issuecomment-2648824198:

<img width="1272" alt="Image" src="https://github.com/user-attachments/assets/c3b31c28-bee3-409f-acf6-cbb45cf4a406" />
The output after adding the two lines of code in train.py is shown above.

**Comment by ddd9898** [comment]
I tried to remove those datasets with too big size, but still get this error:

<img width="1093" alt="Image" src="https://github.com/user-attachments/assets/98e7c4e3-3af8-42c9-a5af-e101dc9396bb" />

**Comment by ddd9898** [comment]
When running train_2D.py I enabled faulthandler to allow Python to print out stack information when the program crashes, 

<img width="512" alt="Image" src="https://github.com/user-attachments/assets/a9864d62-04cd-429b-a3e7-cac4e366e8ee" />

and the error is reported in more detail below:
 
>Training:   0%|          | 0/1000 [00:00<?, ?it/s]terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
Fatal Python error: Aborted

>Thread 0x000040064441f1e0 (most recent call first):
  File "micromamba/envs/csc/lib/python3.11/site-packages/xarray_tensorstore.py", line 135 in __array__
  File "micromamba/envs/csc/lib/python3.11/site-packages/xarray_tensorstore.py", line 139 in get_duck_array
  File "micromamba/envs/csc/lib/python3.11/site-packages/xarray/core/variable.py", line 415 in data
  File "micromamba/envs/csc/lib/python3.11/site-packages/xarray/core/dataset.py", line 4183 in interp
  File "micromamba/envs/csc/lib/python3.11/site-packages/xarray/core/dataarray.py", line 2378 in interp
  File "micromamba/envs/csc/lib/python3.11/site-packages/cellmap_data/image.py", line 508 in return_data
  File "micromamba/envs/csc/lib/python3.11/site-packages/cellmap_data/image.py", line 486 in apply_spatial_transforms
  File "micromamba/envs/csc/lib/python3.11/site-packages/cellmap_data/image.py", line 155 in __getitem__
  File "micromamba/envs/csc/lib/python3.11/site-packages/cellmap_data/dataset.py", line 500 in get_label_array
  File "micromamba/envs/csc/lib/python3.11/concurrent/futures/thread.py", line 58 in run
  File "micromamba/envs/csc/lib/python3.11/concurrent/futures/thread.py", line 83 in _worker
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 975 in run
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 1038 in _bootstrap_inner
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 995 in _bootstrap

>Thread 0x000040065aaff1e0 (most recent call first):
  File "micromamba/envs/csc/lib/python3.11/site-packages/xarray_tensorstore.py", line 135 in __array__
  File "micromamba/envs/csc/lib/python3.11/site-packages/xarray_tensorstore.py", line 139 in get_duck_array
  File "micromamba/envs/csc/lib/python3.11/site-packages/xarray/core/variable.py", line 415 in data
  File "micromamba/envs/csc/lib/python3.11/site-packages/xarray/core/dataset.py", line 4183 in interp
  File "micromamba/envs/csc/lib/python3.11/site-packages/xarray/core/dataarray.py", line 2378 in interp
  File "micromamba/envs/csc/lib/python3.11/site-packages/cellmap_data/image.py", line 508 in return_data
  File "micromamba/envs/csc/lib/python3.11/site-packages/cellmap_data/image.py", line 486 in apply_spatial_transforms
  File "micromamba/envs/csc/lib/python3.11/site-packages/cellmap_data/image.py", line 155 in __getitem__
  File "micromamba/envs/csc/lib/python3.11/site-packages/cellmap_data/dataset.py", line 500 in get_label_array
  File "micromamba/envs/csc/lib/python3.11/concurrent/futures/thread.py", line 58 in run
  File "micromamba/envs/csc/lib/python3.11/concurrent/futures/thread.py", line 83 in _worker
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 975 in run
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 1038 in _bootstrap_inner
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 995 in _bootstrap

>Thread 0x00004002b3bcf1e0 (most recent call first):
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 320 in wait
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 622 in wait
  File "micromamba/envs/csc/lib/python3.11/concurrent/futures/_base.py", line 243 in as_completed
  File "micromamba/envs/csc/lib/python3.11/site-packages/cellmap_data/dataset.py", line 511 in get_target_array
  File "micromamba/envs/csc/lib/python3.11/concurrent/futures/thread.py", line 58 in run
  File "micromamba/envs/csc/lib/python3.11/concurrent/futures/thread.py", line 83 in _worker
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 975 in run
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 1038 in _bootstrap_inner
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 995 in _bootstrap

>Thread 0x000040065ad0f1e0 (most recent call first):
  File "micromamba/envs/csc/lib/python3.11/site-packages/xarray_tensorstore.py", line 135 in __array__
  File "micromamba/envs/csc/lib/python3.11/site-packages/xarray_tensorstore.py", line 139 in get_duck_array
  File "micromamba/envs/csc/lib/python3.11/site-packages/xarray/core/variable.py", line 415 in data
  File "micromamba/envs/csc/lib/python3.11/site-packages/xarray/core/dataset.py", line 4183 in interp
  File "micromamba/envs/csc/lib/python3.11/site-packages/xarray/core/dataarray.py", line 2378 in interp
  File "micromamba/envs/csc/lib/python3.11/site-packages/cellmap_data/image.py", line 508 in return_data
  File "micromamba/envs/csc/lib/python3.11/site-packages/cellmap_data/image.py", line 486 in apply_spatial_transforms
  File "micromamba/envs/csc/lib/python3.11/site-packages/cellmap_data/image.py", line 155 in __getitem__
  File "micromamba/envs/csc/lib/python3.11/site-packages/cellmap_data/dataset.py", line 453 in get_input_array
  File "micromamba/envs/csc/lib/python3.11/concurrent/futures/thread.py", line 58 in run
  File "micromamba/envs/csc/lib/python3.11/concurrent/futures/thread.py", line 83 in _worker
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 975 in run
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 1038 in _bootstrap_inner
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 995 in _bootstrap

>Thread 0x0000400216c0f1e0 (most recent call first):
  File "micromamba/envs/csc/lib/python3.11/selectors.py", line 415 in select
  File "micromamba/envs/csc/lib/python3.11/multiprocessing/connection.py", line 930 in wait
  File "micromamba/envs/csc/lib/python3.11/multiprocessing/connection.py", line 423 in _poll
  File "micromamba/envs/csc/lib/python3.11/multiprocessing/connection.py", line 256 in poll
  File "micromamba/envs/csc/lib/python3.11/multiprocessing/queues.py", line 113 in get
  File "micromamba/envs/csc/lib/python3.11/site-packages/tensorboardX/event_file_writer.py", line 202 in run
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 1038 in _bootstrap_inner
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 995 in _bootstrap

>Thread 0x000040021ebbf1e0 (most recent call first):
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 324 in wait
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 622 in wait
  File "micromamba/envs/csc/lib/python3.11/site-packages/tqdm/_monitor.py", line 60 in run
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 1038 in _bootstrap_inner
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 995 in _bootstrap

>Thread 0x0000400079bc4900 (most recent call first):
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 320 in wait
  File "micromamba/envs/csc/lib/python3.11/threading.py", line 622 in wait
  File "micromamba/envs/csc/lib/python3.11/concurrent/futures/_base.py", line 243 in as_completed
  File "micromamba/envs/csc/lib/python3.11/site-packages/cellmap_data/dataset.py", line 559 in __getitem__
  File "micromamba/envs/csc/lib/python3.11/site-packages/torch/utils/data/dataset.py", line 335 in __getitem__
  File "micromamba/envs/csc/lib/python3.11/site-packages/torch/utils/data/_utils/fetch.py", line 51 in <listcomp>
  File "micromamba/envs/csc/lib/python3.11/site-packages/torch/utils/data/_utils/fetch.py", line 51 in fetch
  File "micromamba/envs/csc/lib/python3.11/site-packages/torch/utils/data/dataloader.py", line 675 in _next_data
  File "micromamba/envs/csc/lib/python3.11/site-packages/torch/utils/data/dataloader.py", line 631 in __next__
  File "cellmap/cellmap-segmentation-challenge-main/src/cellmap_segmentation_challenge/train.py", line 368 in train
  File "cellmap/cellmap-segmentation-challenge-main/examples/train_2D.py", line 84 in \<module\>

The bug seems to be related to the dataloader of cellmap_data (https://github.com/janelia-cellmap/cellmap-data).

**Comment by paperplane03** [comment]
I wonders whether this length list is right, which means the largest crop includes over 3e8 samples in dataloader?

**Comment by ddd9898** [comment]
Hi @paperplane03 , could I contact you? Wechat and email are both fine if it's convenient for you.

**Comment by paperplane03** [comment]
you can contact me through 211300069@smail.nju.edu.cn

**Comment by ddd9898** [comment]
I tried downgrading xarray_tensorstore to 0.1.0 (previously it was 0.1.5) and the error disappeared 🤦.

<img width="1031" alt="Image" src="https://github.com/user-attachments/assets/82c84760-48d2-4a0d-9d53-8c1e52d9abcf" />

**Comment by rhoadesScholar** [comment]
> I wonders whether this length list is right, which means the largest crop includes over 3e8 samples in dataloader? [@rhoadesScholar](https://github.com/rhoadesScholar)

That is correct. Every pixel that represents a valid ROI center in a crop constitutes a "unique" sample. This gives the advantage of determinism while maintaining maximum data diversity (i.e. samples can come from randomly overlapping patches).

**Comment by rhoadesScholar** [comment]
> I tried downgrading xarray_tensorstore to 0.1.0 (previously it was 0.1.5) and the error disappeared 🤦.
> 
> <img alt="Image" width="1031" src="https://private-user-images.githubusercontent.com/74239179/433399464-82c84760-48d2-4a0d-9d53-8c1e52d9abcf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ4MjAxMzMsIm5iZiI6MTc0NDgxOTgzMywicGF0aCI6Ii83NDIzOTE3OS80MzMzOTk0NjQtODJjODQ3NjAtNDhkMi00YTBkLTlkNTMtOGMxZTUyZDlhYmNmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE2VDE2MTAzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRkODk3NzRjOTNiZmNiMTA3MWMxNzE5YTBkZTdlMDdmMGFlMzQzZTkwZTZmZjAwY2ZhMjVjNTgwMTA5ZjM5YmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XWizl1EYXX611sTUGKkuSebRkCIsPV6Q2HGJpzLLepY">

Good find! Unfortunately this removes the writing functionality of xarray_tensorstore, which you will need during the submission phase. 

FWIW, I just ran the `train_2D.py` example without a problem with `xarray==2024.10.0` and `xarray-tensorstore==0.1.5`...

**Comment by rhoadesScholar** [comment]
> That also works for me! I remind the first time I run the contest (on Dec 21, 2024) there is no problem, but later problems happen. and finally this is the solution. Maybe we can add the constrain in requirement.txt?

To clarify @paperplane03, the versions I posted worked for you?

**Comment by paperplane03** [comment]
Yes, 'xarray==2024.10.0' and 'xarray-tensorstore==0.1.5' works for me, and the former is the key point.

**Comment by rhoadesScholar** [comment]
Thanks @paperplane03 and @ddd9898! Looks like we found the solution.
